### DISCUSSION TITLE
CSS Reset

### DISCUSSION THREAD

**Discussion by vitor-mendrone-deel** [frontendbr/forum#2] [Discussão]
Pessoal, sei que esse assunto já foi amplamente debatido em centenas de locais mas acredito que essa seja uma discussão valida pra termos por aqui. 

Como vocês tem 'resetado' o CSS? 

Sempre que vou iniciar um novo projeto fico com essa dúvida. Normalize, Reset.css, resetar só o que preciso, resetar tudo (o famoso: \* {margin: 0; Padding: 0; box-sizing: border-box;})

**Comment by matheuslc** [comment]
Eu utilizo o normalize com box-sizing: border-box.

Como dito no post do [Paul Irish](http://www.paulirish.com/2012/box-sizing-border-box-ftw/), isso não vai afetar sua performance.

**Comment by diogomoretti** [comment]
mesmo que o @matheuslc, utilizo [normalize.styl](https://github.com/frontinfloripa/frontinfloripa.com.br/blob/master/src/stylus/_vendor/normalize.styl) e também `box-sizing: border-box` :ok_hand:

**Comment by joaopedron** [comment]
Utilizo o Reset do Eric e também box-sizing: border-box

**Comment by eduardojmatos** [comment]
Uso quase o mesmo esquema seu @Mendrone e do @diogomoretti (normalize + `box-sizing: border-box`)

**Comment by heloisabiagi** [comment]
Usava o reset do Eric Meyer, mas ultimamente tenho dado preferência ao Normalize e ajustando também o box-sizing. O bom do Normalize é que ele não é um reset que deixa todos os elementos "pelados", ele normaliza uma aparência decente em todos os browsers, e isso nos faz poupar um certo tempo.

Quanto ao reset com "*"... eu sei que ele é largamente popular e utilizado, só que a renderização dele é bem mais lenta porque ele vai ser aplicado a todos os elementos do DOM sem exceção. Pra quem precisa de performance, é melhor declarar explicitamente os elementos que vc quer "zerar" (às vezes vc não quer mexer numa checkbox ou radio-button, mas o código vai passar por eles mesmo assim).

**Comment by brunopulis** [comment]
Eu utilizo o Normalize.css :)

**Comment by vitor-mendrone-deel** [comment]
A algum tempo eu encontrei [esse](http://jaydenseric.com/blog/forget-normalize-or-resets-lay-your-own-css-foundation) reset, bastante simples ainda mas com alguns pontos interessantes: 

``` css
body {
    margin: 0;
    font: 16px/1 sans-serif;
    -moz-osx-font-smoothing: grayscale;
    -webkit-font-smoothing: antialiased;
}
```

Que garante maior consistência no visual de fontes (principalmente entre OsX e Windows

Mas ainda tenho utilizado somente normalize + box-sizing

**Comment by eduardojmatos** [comment]
`margin: 0;` é uma boa também...

Não sabia do `-moz-osx-font-smoothing`, mas o foda é que é non-standard... já já Firefox pode remover.

**Comment by fdaciuk** [comment]
+1 pro Normalize + box-sizing e umas pequenas modificações.. essa parada do font-smoothing eu uso tbem xD

Mas isso pra projetos de médio e grande porte.. pra projetos pequenos, não dá tanta diferença, então é de boa pra usar o polêmico `* { border: 0; margin: 0; padding: 0 }`.

xD

**Comment by jacksonfdam** [comment]
Acho valido olhar essa [thread](http://stackoverflow.com/questions/11578819/css-reset-what-exactly-does-it-do)!

**Comment by eduardojmatos** [comment]
@jacksonfdam bacana, citaram o http://cssreset.com

esse artigo http://sixrevisions.com/css/should-you-reset-your-css/ é bem longo, lerei com calma mais tarde pra avaliar heheh

**Comment by cristianounix** [comment]
Eu tb faço o mesmo do @fdaciuk 
+normalize +box-sizing

**Comment by marquinhusgoncalves-zz** [comment]
Também utilizo o Normalize com box-sizing: border-box

**Comment by tcelestino** [comment]
Tem alguém que não usa nenhuma dessas opções? Que prefere resetar a partir das necessidades?

**Comment by felipecodes** [comment]
Eu prefiro o CSS Reset do Eric Meyer. Já usei o Normalize, mas não curti muito porque eu sempre tinha que resetar os elementos que ele não reseta.

**Comment by vitor-mendrone-deel** [comment]
Aparentemente Normalize + box-sizing FTW

O Reset.css nem foi citado ainda.

**Comment by vitor-mendrone-deel** [comment]
Ok, agora foi. HAHAHAH

**Comment by dleitee** [comment]
+1 para Normalize com box-sizing: border-box

**Comment by napalmdev** [comment]
+1 Normalize + box-sizing: border-box

**Comment by bernardodiasc** [comment]
lembrando que normalize não reseta o css. resetar o css é tirar as propriedades de apresentação, algo que o normalize não faz. o que ele faz é deixar a aprensentação o mais identica possivel em todos os browsers.

bom, até mesmo o reset do eric meyer não limpa toda a aprensentação padrão que os browsers usam nos elementos, por exemplo font-weight dos headers (h1, h2, h3 ...) fica o definido pelo browser, ele não reseta isso. só exemplificando mesmo. ele limpa as principais propriedades que usamos pra layout, por exemplo margin e padding, font-size, etc.

um reset completo removeria até font-weight dos headers, na minha humilde opinião. digo, não importa como o browser quer mostrar os elementos por padrão, se o design do projeto for diferente, aquele estilo vai ser sopreposto de qualquer forma, não preciso do que o browser fornece, melhor pra mim se souber que ta tudo zerado na hora de começar. outro motivo é que as vezes, especialmente quando somos iniciantes no ramo, confundimos o que vemos com o que as coisas realmente são, mas isso é assunto pra outra discussão, sobre semantica e html (css sem html serve pra que?).

:)

**Comment by diogomoretti** [comment]
Concordo @bernardodiasc, apesar da nomenclatura não ser a mais apropriada, a comunidade também categoriza/chama o normalize de CSS Reset :)

**Comment by matheuslc** [comment]
@bernardodiasc Na verdade o que um CSS Reset, é "resetar" para uma padrão algumas propriedades que os browsers definem com estilos diferentes. Não é preciso re-estilizar os headers já que todos os browsers dão o mesmo font-weight para eles por padrão.

**Comment by bernardodiasc** [comment]
Bem observado, isso eh o que o normalize faz, acho que outros resets sao diferente. O normalize tem esse nome exatamente pra direcionar melhor a terminologia. Normalizar seria o que estamos chamando de reiniciar, certo? Reiniciar o estilo/aparencia de um bloco de texto significa, tirar tudo que o browser define por padrao, o h1 fica igual p. Normalizar seria corrigir as diferenças dos browsers, era a proposta dos reset css das antigas, mas mal aplicada, pq eh muito mais facil resetar do que normalizar. Da pra fazer um site ate elegante soh com normalize, soh com css reset nao fica bom. Posso estar enganado, to lonje do pc pra testar, mas eh facil imaginar, soh ver as margens em bloco de texto, se tirar isso fica bem ruim de ler, o normalize nao faz esse tipo de coisa.

Essas foram as razoes de escolher normalize em todos os projetos meus, e um ou outro algum tipo de reset com propriedades especificas do projeto (ou inves de remover tudo, definir as bases do projeto, tipo isso).

**Comment by felipefialho** [comment]
Caras, e uma solução mais simples como mostrada pelo @eduardojmatos?

Eu cheguei a procurar algo assim uma época, algo simples mesmo, quase como o velho \* {margin 0 padding 0 }, mas estou acostumado com o normalize que não fui afundo nisso.

Eu costumo dar uma limpada no arquivo e deixar só blocos que vou utilizar no projeto, mas mesmo assim, as vezes considero um pouco canhão demais.

**Comment by vitor-mendrone-deel** [comment]
É esse meu eterno problema @LFeh, sempre acho que o Normalize é um canhão pra problemas pequenos, mas nunca tenho tempo de procurar um outro caminho.

**Comment by matheuslc** [comment]
Eu acho bem pequeno na real. Acho que tem que ser um projeto MUITO específico para justificar o não gasto de kbs.

**Comment by abnersajr** [comment]
Eu tenho uma versão que eu limpei com a maioria das coisas que não costumo usar.  Trabalho em agência, mutias vezes nem table uso.
Eu vou criar um repositório com uma versão Lite, ajustada as minhas necessidades e explicar o por que da remoção de cada trecho.
Porém como o @matheuslc falou, muitas vezes importamos no começo do projeto, usando um bower ou um pré-processador e acredito que não venha a ser um "canhão" tão grande a ponto de impactar MUITO um projeto.

**Comment by kazzkiq** [comment]
Eu geralmente vou de Normalize.css, mas quando por algum motivo não posso usá-lo (projeto legado por exemplo, onde o Normalize gera resultados imprevisíveis em todas as telas) geralmente opto por algo mais simples:

``` css
body {
    margin: 0;
    min-height: 100vh;
}

*,
*::before,
*::after {
    box-sizing: border-box;
}

*:focus {
    outline: none;
}

a {
    color: inherit;
}
```

Já dá uma facilitada e é mais raro de quebrar projetos que já tem seu próprio CSS, etc.

**Comment by idmarjr** [comment]
Durante alguns bons anos usei o [HTML5Reset do HTML5Doctor](http://html5doctor.com/html-5-reset-stylesheet/). 

Faz um ano resolvi usar o código dele como base e complementar com algumas coisas que sempre surgiam ao longo dos projetos, daí criei o [Aerochimps Reset CSS](https://github.com/Aerochimps/acResetCSS/blob/master/acReset.css).

As minhas adições [começam na linha 121](https://github.com/Aerochimps/acResetCSS/blob/master/acReset.css#L121).

---

O objetivo do [acResetCSS](https://github.com/Aerochimps/acResetCSS/blob/master/acReset.css) é ser um **reset** de verdade e não um "estabilizador" como o Normalize.css. Particularmente prefiro começar o projeto com o máximo de estilos zerados pra ter mais controle sobre o código e deixar mais evidente quando alguma parte do software não recebeu atenção nos estilos. É claro que isso acaba dando um pouco mais de trabalho mas resulta em uma estrutura muito mais confiável, na minha opinião.

**Comment by vitor-mendrone-deel** [comment]
Ja usei algumas vezes uma solução parecida com essa @kazzkiq, só não removo o outline dos elementos por questões de acessibilidade.

**Comment by raphaelfabeni** [comment]
_Normalize_ era padrão da maioria dos projetos que mexia. Atualmente para projetos menores reparei que acabei fazendo _na mão_. Mas para coisas maiores, acho que voltaria a recorrer a algum similar ao carinha _Normalize_.

**Comment by designernatan** [comment]
O Normalize me atende bem e sempre o utilizo nos meus jobs.

**Comment by ghost** [comment]
@kazzkiq tenho essa pegada de fazer esse simples reset, manual mesmo, não descarto voltar para os resets, mas hoje está  tranquilo o manual :)

**Comment by wschenkel** [comment]
Para projetos pequenos acabo usando o `* { box-sizing: border-box; margin: 0; padding: 0; }` mas para projetos maiores opto pelo normalize.

**Comment by elvisdetona** [comment]
Hj em dia como 99,9% dos projetos usamos algum preprocessor, acho melhor zerar na mão somente o que preciso. Sempre usei o normalize + `box-sizing: border-box`  (assim que descobri), mas HJ, 2016, prefiro zerar só o que preciso. 
Eu trabalho com muiiiiiiiiito codigo legado, então zerar só o que preciso é sempre melhor e corro menos risco de dar merda.

**Comment by eduardojmatos** [comment]
@elvisdetona então, em legados realmente é outro cenário... Também tô contigo, não arrisco aplicar nem reset nem normalize hehehe

**Comment by kazzkiq** [comment]
@Mendrone Eu geralmente re-estilizo o estado de `:focus` com algum outro tipo de efeito visual, esse "reset" é só pra tirar aquela borda feia dos elementos que me incomoda demais haha (TOC feelings).

@meche Eu sempre que posso ainda uso o Normalize.css, mais por costume do que por necessidade, pra ser sincero. (acho que hoje em dia a renderização nos browsers já está mais homogênea comparado à quando eu comecei a usar)

**Comment by willianjusten** [comment]
Eu prefiro aplicar reset só no que eu preciso, assim evito de criar linhas e linhas de código desnecessárias, podendo gerar um legado grande e perigoso.

**Comment by matheuslc** [comment]
@willianjusten Tu não acha que um reset é simples de mais pra deixar um legado grande e perigoso?

**Comment by felipefialho** [comment]
@kazzkiq Cara, não fala isso =(

Ao tirar o :focus dos elementos você praticamente acaba com a acessibilidade do projeto. Navegação por teclado fica praticamente impossível.

c/c @deividmarques

**Comment by willianjusten** [comment]
@matheuslc , o reset nada mais é que um overwrite de características, e se eu precisar mudar algo, vou ter que reescrever novamente. Se eu faço uma só vez, já deixo resolvido =)

Mas isso varia muito de projeto para projeto, eu sou afixionado por performance...

**Comment by deividmarques** [comment]
@kazzkiq Em projetos web raramente pensa-se  na acessibilidade e ela possui várias camadas ou fases. Antes mesmo de chegar em leitores de tela (usando wai-aria), precisa-se pensar em navegacão com o teclado e como se sabe aonde o foco está visualmente ? 

```
*:focus {
    outline: none;
}
```

O código acima acaba com isso, mas claro que pode-ser fazer isso, no entanto terá um grande trabalho pra adicionar visualmente o comportamento de :focus em todos elementos de formulários, links e botões do seu projeto. 

O twitter tbm zera o outline, no entanto ele mostra ao usuário o foco do teclado através do CSS em cada elemento. Faça você mesmo esse teste no twitter: Vá ao topo, primeiro navegue com o mouse e depois com o teclado utilizando a tecla TAB e entenderás melhor o que citei.
https://twitter.com

**Comment by kazzkiq** [comment]
Opa opa, calma lá pessoal. @LFeh @deividmarques  eu estou ciente do problema de acessibilidade/navegação ao remover os outlines, e justamente por isso que eu afirmei na minha outra resposta **que substituo o outline por algum outro tipo de efeito visual**.

Geralmente substituo outline por bordas + box-shadow (seguindo a identidade visual da aplicação), mas obviamente não coloquei estes estilos substitutos lá no código porque aí ele deixaria de ser um reset. ;)

**Comment by felipefialho** [comment]
Galera.

Vou refazer a questão do @Mendrone... :)

> Vocês ainda estão resetando seus elementos?

**Comment by wschenkel** [comment]
De um tempo para cá, tenho usado esse reset: http://jgthms.com/minireset.css/ 
Tem sido bastante útil.

**Comment by vitor-mendrone-deel** [comment]
Quase 1 ano depois posso dizer que abandonei o Normalize e tenho feito o reset de acordo com as necessidades dos projetos

O minireset citado pelo @willianschenkel é bem legal, utilizei em um projeto pessoal, mas removi o reset de tables que eu uso MUITO raramente.

**Comment by felipefialho** [comment]
Sendo muito sincero, desse tempo pra cá que o @Mendrone criou a issue estou usando um simples:

```
*
  box-sizing border-box
  margin 0
  padding 0 
  
  &:before
  &:after
    box-sizing border-box
```

Isso porque, em basicamente todos os projetos, preciso desse reset básico, mas tem casos (tipo meu site), que reseto só o `box-sizing`!

E complementando de acordo com a especificidade do projeto, por exemplo, no que estou trabalhando atualmente.

```
input[type="search"]
  appearance searchfield

  &::-webkit-search-cancel-button
    appearance searchfield-cancel-button
```

Tenho trabalhado mais com os resets dentro do próprio escopo do elemento.

**Comment by lcquadros** [comment]
Normalize com box sizing é muito bom.

Mas uma vez que ele "reseta" algumas declarações, o melhor que pudemos fazer com ele foi adicionar variáveis default no sass para que fosse mais um reboot do que um reset e - na época - reduzir volume de código.
